File: code/library/src/main/java/org/fbme/lib/iec61499/stringify/ParameterDeclarationPrinter.java
package org.fbme.lib.iec61499.stringify;


import org.fbme.lib.iec61499.declarations.ParameterDeclaration;
import org.fbme.lib.st.expressions.Literal;
import org.jdom.Element;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;

public class ParameterDeclarationPrinter extends DeclarationPrinterBase<ParameterDeclaration> {

    public ParameterDeclarationPrinter(ParameterDeclaration declaration) {
        super(declaration, "VarDeclaration");
    }

    @Override
    protected void printDeclarationBody(Element element) {
        element.setAttribute("Type", Objects.requireNonNull(myElement.getType()).stringify());
        Literal<?> initialValue = myElement.getInitialValue();
        if (initialValue != null) {
            element.setAttribute("InitialValue", STPrinter.printLiteral(initialValue));
        }
    }

    public static @Nullable Element printAll(String collectionTag, List<ParameterDeclaration> declarations) {
        if (declarations.isEmpty()) {
            return null;
        }
        Element element = new Element(collectionTag);
        printAll(declarations, element);
        return element;
    }

    public static void printAll(List<ParameterDeclaration> declarations, Element element) {
        for (ParameterDeclaration declaration : declarations) {
            element.addContent(new ParameterDeclarationPrinter(declaration).print());
        }
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/stringify/ParameterDeclarationPrinter.kt
package org.fbme.lib.iec61499.stringify

import org.fbme.lib.iec61499.declarations.ParameterDeclaration
import org.fbme.lib.st.types.ArrayType
import org.jdom.Element

class ParameterDeclarationPrinter(declaration: ParameterDeclaration) :
    DeclarationPrinterBase<ParameterDeclaration>(declaration, "VarDeclaration") {
    override fun printDeclarationBody(element: Element) {
        val type = this.element.type
        if (type is ArrayType) {
            type.dimensions?.stringify()?.let { element.setAttribute("ArraySize", it) }
            type.baseType?.stringify()?.let { element.setAttribute("Type", it) }
        } else {
            type?.stringify()?.let { element.setAttribute("Type", it) }
        }
        val initialValue = this.element.initialValue
        if (initialValue != null) {
            element.setAttribute("InitialValue", STPrinter.printLiteral(initialValue))
        }
    }

    companion object {
        fun printAll(collectionTag: String?, declarations: List<ParameterDeclaration>): Element? {
            if (declarations.isEmpty()) {
                return null
            }
            val element = Element(collectionTag)
            printAll(declarations, element)
            return element
        }

        @JvmStatic
        fun printAll(declarations: List<ParameterDeclaration>, element: Element) {
            for (declaration in declarations) {
                element.addContent(ParameterDeclarationPrinter(declaration).print())
            }
        }
    }
}
